helix-core/src/history.rs

```diff
@@ -23,72 +64,259 @@ impl Default for History {
+    fn revision_closer_to_instant(&self, i: usize, instant: Instant) -> usize {
+        let dur_im1 = instant.duration_since(self.revisions[i - 1].timestamp);
+        let dur_i = self.revisions[i].timestamp.duration_since(instant);
+        use std::cmp::Ordering::*;
+        match dur_im1.cmp(&dur_i) {
+            Less => i - 1,
+            Equal | Greater => i,
+        }
+    }
+
+    fn jump_instant(&mut self, instant: Instant) -> Vec<Transaction> {
+        let search_result = self
+            .revisions
+            .binary_search_by(|rev| rev.timestamp.cmp(&instant));
+        let revision = match search_result {
+            Ok(revision) => revision,
+            Err(insert_point) => match insert_point {
+                0 => 0,
+                n if n == self.revisions.len() => n - 1,
+                i => self.revision_closer_to_instant(i, instant),
+            },
+        };
+        self.jump_to(revision)
+    }
+
+    fn jump_duration_backward(&mut self, duration: Duration) -> Vec<Transaction> {
+        match self.revisions[self.current].timestamp.checked_sub(duration) {
+            Some(instant) => self.jump_instant(instant),
+            None => self.jump_to(0),
+        }
+    }
+
+    fn jump_duration_forward(&mut self, duration: Duration) -> Vec<Transaction> {
+        match self.revisions[self.current].timestamp.checked_add(duration) {
+            Some(instant) => self.jump_instant(instant),
+            None => self.jump_to(self.revisions.len() - 1),
+        }
+    }
+
+    pub fn earlier(&mut self, uk: UndoKind) -> Vec<Transaction> {
+        use UndoKind::*;
+        match uk {
+            Steps(n) => self.jump_backward(n),
+            TimePeriod(d) => self.jump_duration_backward(d),
+        }
+    }
+
+    pub fn later(&mut self, uk: UndoKind) -> Vec<Transaction> {
+        use UndoKind::*;
+        match uk {
+            Steps(n) => self.jump_forward(n),
+            TimePeriod(d) => self.jump_duration_forward(d),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UndoKind {
+    Steps(usize),
+    TimePeriod(std::time::Duration),
+}
+
+const TIME_UNITS: &[&str] = &["seconds", "minutes", "hours", "days"];
+
+fn find_time_unit(s: &str) -> Result<&'static str, String> {
+    for unit in TIME_UNITS {
+        if unit.starts_with(&s.to_lowercase()) {
+            return Ok(unit);
+        }
+    }
+    Err(format!("incorrect time unit: {}", s))
+}
+
+static DURATION_VALIDATION_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^(?:\d+\s*[a-zA-Z]+\s*)+$").unwrap());
+
+static NUMBER_UNIT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"(\d+)\s*([a-zA-Z]+)").unwrap());
+
+fn parse_human_duration(s: &str) -> Result<Duration, String> {
+    if !DURATION_VALIDATION_REGEX.is_match(s) {
+        return Err("duration should be composed \
+        of positive integers followed by time units"
+            .to_string());
+    }
+
+    let mut units = std::collections::HashMap::new();
```

Comment:
Since this is simple, I think we can just use `[usize; 4]` rather than `HashMap`, no need extra hashing.

Reply:
The approaches are for all intents and purposes equivalent. Which one to use is a matter of personal preference, and it's not worth rewriting one to another.

Reply:
Yeah, but with array we can be sure we have 4 of them. Either way the size is the same as hashmap and we don't have to point to the heap for this. And we also don't have to do `unwrap_or` later on since it will always be there. This is only possible as we know we have a fixed set of time units we support here.

Reply:
Using the heap is really not an issue. It's useful to us to have the option, because then we can detect duplicate use of the same unit (e.g "1s 2s 3s" is not a valid input). We could use two arrays, one with bools or an array of Option<u64>, but this will complicate the code.

Since I'm not attached to this implementation, maybe you could change it after it's upstream? It's easier to think about it when I can see the code change that you have in mind.